### DISCUSSION TITLE
Bug or faulty implementation by me?

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2636] [Evennia Questions] [answered]
I recently implemented an "identify" command for reporting the attributes of equipment to players. I set up both "id" and "iden" as aliases for it. While testing it, I inadvertently typed "ide jersey". This threw an actual Python error with traceback, etc., instead of saying something like "command not found", or something similar, which is what I would have expected. I do not have the error handy, but can certainly replicate and re-obtain.

Certainly I can work around this by just aliasing identify (and any similar command) as "id", "ide", "iden", etc., but, again, I would have thought that this would have gotten a response from the parser/handler of "command not found" or something similar. Is this functioning as intended, or do you want me to gather some more info?

Thanks, and still loving this! So much easier than doing Envy in C!!!!!

**Comment by aMiss-aWry** [comment, marked as answer]
Try going into your settings.py and adding this line: `COMMAND_DEFAULT_ARG_REGEX = r'^[ /]+.*$|$'`

It happens because the base regex at the moment interprets commands that look like 'command + some text' as being that command regardless of spaces involved, so it's trying to parse 'id e jersey' when you typo id/iden as ide. I believe Griatch has updated evennia's 'develop' branch to have that same regex in the next evennia release. Note that this will still allow you to use / for a switch in commands (i.e. typeclass/force).

**Reply by RandomEC** [reply]
Will do, thanks!

**Reply by Griatch** [reply]
@RandomEC Note that, regardless of the arg-regex, a misidentified command should still not raise a traceback. 
Most likely your command failed when receiving a space in the argument or similar - but we'd need to see the full tracecback to be sure.

**Reply by RandomEC** [reply]
@Griatch I think that this is correct. My recollection is that it recognized "id" as a command, but had a problem with "e jersey". I will replicate tonight and post the traceback in a reply. Thanks again.

**Reply by RandomEC** [reply]
@Griatch Here you go!
```python
Could not find 'e jersey'.
Traceback (most recent call last):
 File "c:\users\bradm\mudstuff\evennia\evennia\commands\cmdhandler.py", line 644, in _run_command
   ret = cmd.func()
 File ".\commands\equipment_commands.py", line 296, in func
   name = item.name
AttributeError: 'NoneType' object has no attribute 'name'

An untrapped error occurred.
(Traceback was logged 22-01-25 19:41:29-05).
```

**Reply by Griatch** [reply]
Yes, as you can see, this is a bug in your code. 

> Could not find 'e jersey' 

comes from the standard `caller.search()` method. It gives an error and returns `None` when it finds no matches (or a multimatch). At this point the user has already been properly informed (like you were above) so you should abort and not continue. This command will fail this way whenever you are trying to id something and can't find it. Fix it this way: 

```python
# ...
item = self.caller.search(...)
if not item:
    # if there is no item, the search helper has already told the caller all about it.
    return 
# ...
```

**Reply by RandomEC** [reply]
Thanks! That's normally code that I have in my commands, but somehow I omitted it here. 

I think that I will probably do a combination of this and the iterative aliasing ("id", "ide", "iden", "ident", etc.) I discussed above. I think that its going to be confusing to players to tell them, essentially, that there is a problem with the object that they are trying to find when the real problem is that they are not inputting the command correctly. In other words, the problem isn't that they tried to look for "e jersey", the problem is that they typed "ide", when only "id" or "iden" will work. 

Thanks again!

**Reply by Griatch** [reply]
Well, for that, the arg-regex change suggested earlier will make sure that "ide" leads to a "command not found" error. :)  You can set `arg_regex` per-command too, if you want.

**Reply by RandomEC** [reply]
Perfect. :) I think that gets it all!
